Perform null check when constructing RawHeader

laszlovandenhoek opened the pull request:
This resolves issue #2103.

jlprat commented on akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala:
```diff
@@ -668,6 +668,8 @@ final case class Range(rangeUnit: RangeUnit, ranges: immutable.Seq[ByteRange]) e
 }
 
 final case class RawHeader(name: String, value: String) extends jm.headers.RawHeader {
+  require(name != null, "name must not be null")
```
@ktoso I remember your comment about `require` allocating a closure in another PR. So I guess we should use here as well `if(name == null) throw ...`.

johanandren replied:
True here as well, please if-throw instead.

ktoso replied:
Ah good catch @jlprat! thanks! :)

jlprat replied:
It got me by surprise on the other PR, so I tried to remember this one :)

laszlovandenhoek replied:
There's quite a few similar `require` calls in the file. Should those be changed too?

ktoso replied:
Yeah, good idea

johanandren commented:
One thought though, is there a reason RawHeader is specifically likely to get nulls passed, compared to any other header, or any other method in the Java APIs? If not, then shouldn't we either add _everywhere_ or nowhere to be consistent about it?

jlprat commented:
@johanandren I can't imagine why specifically `RawHeader` is more prone to it. One thing that came to my mind as well, is why `null` for the value shouldn't be considered as the empty string? In the end, it represents the absence of value. And I've seen headers that don't have any meaningful value, so they are left empty.

ktoso commented:
> One thing that came to my mind as well, is why null for the value shouldn't be considered as the empty string? In the end, it represents the absence of value. And I've seen headers that don't have any meaningful value, so they are left empty.

-1 for such treatment of nulls; null is wrong and we should throw then. I would suspect if a null ended up there it is a programming error and not an intended thing

ktoso commented:
Raw headers are perhaps a bit more likely to get nulls since people may want to expose "pass through" or "anything i don't understand" on as raw headers, perhaps without making sure the values are valid; I'm ok with the null checks here and not in all other places for that reason

jlprat commented:
>-1 for such treatment of nulls;

Agree, this was my old Java dev in me :laughing:

raboof commented:
@laszlovandenhoek are you still interested in updating the `require`s in this PR? if not I can have a look at it

laszlovandenhoek commented:
@raboof I updated the `require`s. I also added a test (as requested by Konrad) and rebased on `master` (because who likes merge commits).

Actually, I had the fix ready in code, but didn't push it because I had trouble running the tests locally (still do, actually, but that's probably me, so nvm), and subsequently forgot all about it.

Thanks for the nudge!

laszlovandenhoek commented:
Also, annoyingly, [it's its, not it's](http://www.its-not-its.info/).

laszlovandenhoek commented:
PLS BUILD
